group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeSet.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.assignor;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * A {@code RangeSet} represents a range of integers from {@code from} (inclusive)
+ * to {@code to} (exclusive).
+ * This implementation provides a view over a continuous range of integers without actually storing them.
+ */
+class RangeSet implements Set<Integer> {
+    private final int from;
+    private final int to;
+
+    /**
+     * Constructs a {@code RangeSet} with the specified range.
+     *
+     * @param from      The starting value (inclusive) of the range.
+     * @param to        The ending value (exclusive) of the range.
+     */
+    public RangeSet(int from, int to) {
+        this.from = from;
+        this.to = to;
+    }
+
+    @Override
+    public int size() {
+        return to - from;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (o instanceof Integer) {
+            int value = (Integer) o;
+            return value >= from && value < to;
+        }
+        return false;
+    }
+
+    @Override
+    public Iterator<Integer> iterator() {
+        return new Iterator<Integer>() {
+            private int current = from;
+
+            @Override
+            public boolean hasNext() {
+                return current < to;
+            }
+
+            @Override
+            public Integer next() {
+                if (!hasNext()) throw new NoSuchElementException();
+                return current++;
+            }
+        };
+    }
+
+    @Override
+    public Object[] toArray() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
I have noticed that many tests are failing in the build. I suspect that they do because those two methods are not implemented.

Reply:
From the logs: 
```
java.lang.UnsupportedOperationException
	at org.apache.kafka.coordinator.group.assignor.RangeSet.toArray(RangeSet.java:83)
	at java.base/java.util.ArrayList.<init>(ArrayList.java:181)
	at org.apache.kafka.coordinator.group.CoordinatorRecordHelpers.newTargetAssignmentRecord(CoordinatorRecordHelpers.java:242)
	at org.apache.kafka.coordinator.group.consumer.TargetAssignmentBuilder.build(TargetAssignmentBuilder.java:368)
```

Reply:
The `TargetAssignmentBuilderBenchmark` results are very likely impacted by this.